File: dotnet/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example2/HighValueOrderIterator.cs
namespace IteratorDesignPattern.Example2;

public class HighValueOrderIterator : IIterator<Order>
{
    private readonly OrderCollection _collection;
    private readonly decimal _minAmount;
    private int _currentIndex = 0;

    public HighValueOrderIterator(OrderCollection collection, decimal minAmount)
    {
        _collection = collection;
        _minAmount = minAmount;
    }

    public bool HasNext()
    {
        while (_currentIndex < _collection.Count)
        {
            if (_collection[_currentIndex].TotalAmount >= _minAmount)
            {
                return true;
            }
            _currentIndex++;
        }
        return false;
    }

    public Order Next()
    {
        if (!HasNext())
        {
            throw new InvalidOperationException("No more high value orders.");
        }

        return _collection[_currentIndex++];
    }

    public void Reset()
    {
        _currentIndex = 0;
    }
}

File: java/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example2/src/main/java/behavioral/iterator/example2/HighValueOrderIterator.java
package behavioral.iterator.example2;

public class HighValueOrderIterator implements IIterator<Order> {
    private final OrderCollection collection;
    private final double minAmount;
    private int currentIndex = 0;

    public HighValueOrderIterator(OrderCollection collection, double minAmount) {
        this.collection = collection;
        this.minAmount = minAmount;
    }

    @Override
    public boolean hasNext() {
        while (currentIndex < collection.getCount()) {
            if (collection.get(currentIndex).getTotalAmount() >= minAmount) return true;
            currentIndex++;
        }
        return false;
    }

    @Override
    public Order next() {
        if (!hasNext()) throw new java.util.NoSuchElementException();
        return collection.get(currentIndex++);
    }

    @Override
    public void reset() { currentIndex = 0; }
}
